### PR TITLE
APPS/IO-DEMO: Invalidate send IOV when operation completed

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Invalidate send IOV when the operation is completed.

## Why ?

To check if a peer reads (or local side writes) invalid data to a remote buffer which is verified then by a target side.

## How ?

1. Invalidate memory allocated for `BufferIov` during `release()` if it was filled.
2. Some minor code improvements and cleanups.

### CODE REVIEW

File: test/apps/iodemo/io_demo.cc [hunk 1]
```diff
@@ -598,11 +608,18 @@ class P2pDemoCommon : public UcxContext {
         }
 
         void release() {
+            assert(!_iov.empty() || _extra_buf);
+
+            if (_filled) {
```
[the remark this answers is not in the document]
Comment (yosefe) [reply test/apps/iodemo/io_demo.cc:623]: if all buffers are filled, why need the _filled flag? in which case it's false?

**Comment by yosefe** [comment]
error seems relevant:
```
Error iodemo analyzer: io_demo: io_demo.cc:565: size_t P2pDemoCommon::BufferIov::size() const: Assertion `!_iov.empty() || _extra_buf' failed.
```

Maybe make the code beautification/asserts in another PR?

**Comment by dmitrygx** [comment]
> error seems relevant:
> 
> ```
> Error iodemo analyzer: io_demo: io_demo.cc:565: size_t P2pDemoCommon::BufferIov::size() const: Assertion `!_iov.empty() || _extra_buf' failed.
> ```
> 
> Maybe make the code beautification/asserts in another PR?

it was a bug in initial implementation, fixed in the latest commit

**Comment by dmitrygx** [comment]
> > error seems relevant:
> > ```
> > Error iodemo analyzer: io_demo: io_demo.cc:565: size_t P2pDemoCommon::BufferIov::size() const: Assertion `!_iov.empty() || _extra_buf' failed.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Maybe make the code beautification/asserts in another PR?
> 
> it was a bug in initial implementation, fixed in the latest commit

yes, the latest commit solves the issues.
@yosefe could you review pls?

File: test/apps/iodemo/io_demo.cc [hunk 2]
```diff
@@ -591,12 +598,15 @@ class P2pDemoCommon : public UcxContext {
 
             if (validate) {
                 fill_data(sn, conn_id, _memory_type);
+                _filled = true;
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:601]: _validate = validate

Comment (dmitrygx) [reply test/apps/iodemo/io_demo.cc]: done

**Review by yosefe — APPROVED** [review]

**Pull request merged**